pkg/fsm/metrics/internal/processing_duration.go

```diff
@@ -0,0 +1,159 @@
+package internal
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/btree"
+)
+
+type ProcessingStartTimes struct {
+	m          sync.Mutex
+	startTimes *btree.BTreeG[requestStartTime]
+}
+
+// requestStartTime is a wrapper around reconcile.Request that adds a Generation field.
+type requestStartTime struct {
```

Comment:
since its wrapping/importing could we instead define this as:

```golang
type requestStartTime struct {
  Generation int64
  reconcile.Request
}
```

?

I was also thinking if it might be cleaner to have separate structs for the keys/values. WDYT?

Reply:
1. I originally had embedded `reconcile.Request` but this ended up making things a bit harder to read because you have to define the struct with:

```golang
r := requestStartTime{
		Request: reconcile.Request{
			NamespacedName: types.NamespacedName{
				Namespace: namespace,
				Name:      name,
			},
		},
		Generation: observedGeneration,
	}
```

2. I also thought to have separate structs for keys vs values, but came to the same conclusion as above (the extra nested structs made things slightly harder to read). I looked at some other usage examples of BTree (in controller-runtime, for instance) and they just use a single struct to house both key and values.